lib/elixir/pages/mix-and-otp/task-and-gen-tcp.md

```diff
@@ -195,120 +186,120 @@ to also use `Task.start_link/1`:
+#PID<0.48.0>
+iex> Process.exit(pid, :kill)
+true
+iex> Process.whereis(:shopping)
+#PID<0.50.0>
+```
 
-  @doc """
-  Starts accepting connections on the given `port`.
-  """
-  def accept(port) do
-    {:ok, socket} = :gen_tcp.listen(port,
-                      [:binary, packet: :line, active: false, reuseaddr: true])
-    Logger.info "Accepting connections on port #{port}"
-    loop_acceptor(socket)
-  end
+What exactly happens when a process terminates is part of its child specification. For `KV.Bucket`, we have this:
 
-  defp loop_acceptor(socket) do
-    {:ok, client} = :gen_tcp.accept(socket)
-    {:ok, pid} = Task.Supervisor.start_child(KVServer.TaskSupervisor, fn -> serve(client) end)
-    :ok = :gen_tcp.controlling_process(client, pid)
-    loop_acceptor(socket)
-  end
+```elixir
+iex> KV.Bucket.child_spec([])
+%{id: KV.Bucket, start: {KV.Bucket, :start_link, [[]]}}
+```
 
-  defp serve(socket) do
-    socket
-    |> read_line()
-    |> write_line(socket)
+However, for tasks, we have this:
 
-    serve(socket)
-  end
+```elixir
+iex> Task.child_spec(fn -> :ok end)
+%{
+  id: Task,
+  restart: :temporary,
+  start: {Task, :start_link, [#Function<43.39164016/0 in :erl_eval.expr/6>]}
+}
+```
 
-  defp read_line(socket) do
-    {:ok, data} = :gen_tcp.recv(socket, 0)
-    data
-  end
+Notice that a task says `:restart` is `:temporary`. `KV.Bucket` says nothing, which means it defaults to `:permanent`. `:temporary` means that a process is never restarted, regardless of why it crashed. `:permanent` means a process is always restarted, regardless of the exit reason. There is also `:transient`, which means it won't be restarted as long as it terminates successfully.
 
-  defp write_line(line, socket) do
-    :gen_tcp.send(socket, line)
-  end
-end
-```
+Now we must ask ourselves, are those the correct settings?
 
-Since we have changed the supervisor specification, we need to ask: is our supervision strategy still correct?
+For `KV.Bucket`, we could argue that any value works. After all, if a bucket crashes, `KV.locate/1` already has logic for starting a new one in its place dynamically. However, for tasks, we have used them in two opposing ways in this chapter, which means at least one of them is wrong.
 
-In this case, the answer is yes: if the acceptor crashes, there is no need to crash the existing connections. On the other hand, if the task supervisor crashes, there is no need to crash the acceptor too.
+We use a task to start the acceptor. The acceptor is a critical component of our infrastructure. If it crashes, it means we won't accept further requests, and our server would them be useless if no one can connect to it. On the other hand, we also use `Task.Supervisor` to start tasks that deal with each connection. In this case, restarting may not be useful at all, given the reason we crashed could just as well be a connection issue, and attempting to restart over the same connection would lead to further failures.
```

Comment:
```suggestion
We use a task to start the acceptor. The acceptor is a critical component of our infrastructure. If it crashes, it means we won't accept further requests, and our server would then be useless as no one can connect to it. On the other hand, we also use `Task.Supervisor` to start tasks that deal with each connection. In this case, restarting may not be useful at all, given the reason we crashed could just as well be a connection issue, and attempting to restart over the same connection would lead to further failures.
```